src/platforms/stlinkv3/usb_dwc_common.c

```diff
@@ -0,0 +1,480 @@
+ * the adjustments I made. It seems to me that the stm-cube handling of incoming (OUT)
+ * usb packages is more robust, and I decided to keep it this way.
+ *
+ * However, for that to work, it is needed that the packets are read from the usb FIFO packet
+ * memory in some temporary storage, and are handed to the upper usb layers only
+ * when the corresponding XFRC interrupts are received. This means that some temporary
+ * storage is necessary, so this variable serves this purpose.
+ *
+ * Also, note that in usb_control.c, function '_usbd_control_setup()' expects that
+ * the setup data packet (8 bytes) has already been read in the
+ * 'usbd_dev->control_state.req' buffer prior to invoking the '_usbd_control_setup()'
+ * function. I am not very familiar with the usb stack, but it seems more natural
+ * to me that the setup packet is actually read in the '_usbd_control_setup()'
+ * function, instead of handling it as a special case here.
+ */
+static struct incoming_packet {
+	bool is_packet_present;
+	int packet_length;
+	uint8_t packet_data[MAX_BULK_PACKET_SIZE];
+} stashed_packets[USB_ENDPOINT_COUNT];
+
+void dwc_set_address(usbd_device *usbd_dev, uint8_t addr)
+{
+	REBASE(OTG_DCFG) = (REBASE(OTG_DCFG) & ~OTG_DCFG_DAD) | (addr << 4);
+}
+
+void dwc_ep_setup(usbd_device *usbd_dev, uint8_t addr, uint8_t type, uint16_t max_size,
+	void (*callback)(usbd_device *usbd_dev, uint8_t ep))
+{
+	/*
+	 * Configure endpoint address and type. Allocate FIFO memory for
+	 * endpoint. Install callback function.
+	 */
+	uint8_t dir = addr & 0x80;
+	addr &= 0x7f;
+
+	if (addr == 0) { /* For the default control endpoint */
+		/* Configure IN part. */
+		if (max_size >= 64) {
+			REBASE(OTG_DIEPCTL0) = OTG_DIEPCTL0_MPSIZ_64;
+		} else if (max_size >= 32) {
+			REBASE(OTG_DIEPCTL0) = OTG_DIEPCTL0_MPSIZ_32;
+		} else if (max_size >= 16) {
+			REBASE(OTG_DIEPCTL0) = OTG_DIEPCTL0_MPSIZ_16;
+		} else {
+			REBASE(OTG_DIEPCTL0) = OTG_DIEPCTL0_MPSIZ_8;
+		}
+
+		REBASE(OTG_DIEPTSIZ0) = (max_size & OTG_DIEPSIZ0_XFRSIZ_MASK);
+		REBASE(OTG_DIEPCTL0) |=
+			/* OTG_DIEPCTL0_EPENA | */ OTG_DIEPCTL0_SNAK;
+
+		/* Configure OUT part. */
+		usbd_dev->doeptsiz[0] = OTG_DIEPSIZ0_STUPCNT_1 | OTG_DIEPSIZ0_PKTCNT | (max_size & OTG_DIEPSIZ0_XFRSIZ_MASK);
+		REBASE(OTG_DOEPTSIZ(0)) = usbd_dev->doeptsiz[0];
+		REBASE(OTG_DOEPCTL(0)) |= OTG_DOEPCTL0_EPENA | OTG_DIEPCTL0_SNAK;
+
+		REBASE(OTG_GNPTXFSIZ) = ((max_size / 4) << 16) | usbd_dev->driver->rx_fifo_size;
+		usbd_dev->fifo_mem_top += max_size / 4;
+		usbd_dev->fifo_mem_top_ep0 = usbd_dev->fifo_mem_top;
+
+		return;
+	}
+
+	if (dir) {
+		REBASE(OTG_DIEPTXF(addr)) = ((max_size / 4) << 16) | usbd_dev->fifo_mem_top;
+		usbd_dev->fifo_mem_top += max_size / 4;
+
+		REBASE(OTG_DIEPTSIZ(addr)) = (max_size & OTG_DIEPSIZ0_XFRSIZ_MASK);
+		REBASE(OTG_DIEPCTL(addr)) |=
+			/*OTG_DIEPCTL0_EPENA |*/ OTG_DIEPCTL0_SNAK | (type << 18) | OTG_DIEPCTL0_USBAEP | OTG_DIEPCTLX_SD0PID |
+			(addr << 22) | max_size;
+
+		if (callback) {
```

Comment:
Braces [we'll let you clean up the rest of this file for this lint rather than keeping writing it out]